topics/dapp-wallet-msg.md

```diff
@@ -0,0 +1,110 @@
+        - basic information about the wallet (`appMetadata`),
+        - a list of networks on which the user has agreed the relying part can operate, should be a subset of the equivalent from `permission_request` (`networks`),
+        - a list of permission scopes the user has agreed the relying party can be granted, should be a subset of the equivalent from `permission_request` (`scopes`)
+        - the account details (princial or public key (*TBD what is required*) and optional subaccount) of the user selected account (`account`)
+    - If the user rejects the request, the wallet sends `error` back to the relying party with appropriate explanation.
+
+```
+{ // Request
+    type: "permission_request"
+    appMetadata: /* dApp */ {
+        name: String
+        icon: Optional<String>
+    }
+    networks: Array<{
+        type: "mainnet" | "testnet"
+        name: Optional<String>
+        rpcUrl: Optional<String>
+    }>
+    scopes: Array<"canister_call">
+}
+
+{ // Response
+    type: "permission_response"
+    appMetadata: /* wallet */ {
+        name: String
+        icon: Optional<String>
+    }
+    networks: Array<{
+        type: "mainnet" | "testnet"
+        name: Optional<String>
+        rpcUrl: Optional<String>
+    }>
+    scopes: Array<"canister_call">
+    account: {
+        owner: String /* principal / publicKey */
+        subaccount: Optional<String>
+    }
+}
+```
+
+## Canister Call
+
+Once the connection between the relying party and wallet is established, the relying party can request the wallet to execute canister calls:
+1. The relying party sends `canister_call_request`. The message incudes:
+    - the network details on which the call should be executed,
+    - the ID of the target canister,
+    - the name of the call method to be executed,
+    - the arguments for the call.
+2. The wallet receives the request, validates if the relying party has permission to request the action and processes the message (in accordance with the ICRC-21 specification).
+    - If the user approves the request, the wallet sends the call to the IC and, upon receiving a successful response, sends `canister_call_response` back to the relying party. The message includes:
+        - the response from IC,
+        - optionally (*TBD*), the parameters (or their subset) of the request.
+    - If the user rejects the request or wallet fails to complete the requested action for any reason, the wallet sends `error` back to the relying party with appropriate explanation.
+
+```
+{ // Request
+    type: "canister_call_request"
+    network: {
+        type: "mainnet" | "testnet"
+        name: Optional<String>
+        rpcUrl: Optional<String>
+    }
+    canisterId: String
+    method: String
+    args: String /* encoded */ / Object /* decoded */
+}
+
+{ // Response
+    type: "canister_call_response"
+    response: String /* encoded */ / Object /* decoded */
```

Comment:
It would be returned in the response as if the call was successful. 

> I think it would be beneficial to specify that errors originating from the IC should be handed through to the relying party.

Yes, I agree. I think that thanks to there messages the wallet doesn't have to know the service it's calling, hence doesn't have to know how to decode the response. And if it doesn't know how to decode the response, it can't and shouldn't tell if it's successful or not 😄